src/semaphore.c

```diff
@@ -804,3 +880,130 @@ _dispatch_thread_semaphore_wait(_dispatch_thread_semaphore_t sema)
 #error "No supported semaphore type"
 #endif
 }
+
+#if USE_FUTEX_SEM
+#define DISPATCH_FUTEX_NUM_SPINS 100
+#define DISPATCH_FUTEX_VALUE_MAX INT_MAX
+#define DISPATCH_FUTEX_NWAITERS_SHIFT 32
+#define DISPATCH_FUTEX_VALUE_MASK ((1ull << DISPATCH_FUTEX_NWAITERS_SHIFT) - 1)
+
+static int _dispatch_futex_trywait(dispatch_futex_t *dfx);
+static int _dispatch_futex_wait_slow(dispatch_futex_t *dfx,
+		const struct timespec* timeout);
+static int _dispatch_futex_syscall(dispatch_futex_t *dfx, int op, int val,
+		const struct timespec *timeout);
+
+/* all memory orders below were added with strict seq_cst semantics
+ * assumed to be the strongest
+ */
+
+int
+_dispatch_futex_init(dispatch_futex_t *dfx)
+{
+	dfx->dfx_data = 0;
+	return 0;
+}
+
+int
+_dispatch_futex_dispose(dispatch_futex_t *dfx)
+{
+	(void)dfx;
+	return 0;
+}
+
+// Increments semaphore value and reads wait count in a single atomic
+// operation, with release MO. If wait count is nonzero, issues a FUTEX_WAKE.
+int
+_dispatch_futex_signal(dispatch_futex_t *dfx)
+{
+	uint64_t orig;
+	do {
+		orig = dfx->dfx_data;
+		if (slowpath((orig & DISPATCH_FUTEX_VALUE_MASK) ==
+					 DISPATCH_FUTEX_VALUE_MAX)) {
+			DISPATCH_CRASH("semaphore overflow");
+		}
+	} while (!dispatch_atomic_cmpxchg2o(dfx, dfx_data, orig, orig + 1, seq_cst));
```

Comment:
also if you use the union, you would not have to mask anymore, because you'd load into dfx->dfx_data but could read the member which would be significantly more readable